### DISCUSSION TITLE
take returns an Option, when can it be None

### DISCUSSION THREAD

**Discussion by ds-gipe** [blackbeam/mysql_async#306] [General]
I am using the take api and in order to handle nulls I need to do something like this:
`  let qualifying: Option<u32> = if let Some(some_val) = row.take("isQualifying") {
        some_val
    } else {
        None
    };
`

What I don't understand, is the take api returns an Optional. At what case can it return None if not in the null case.

If I try 

`  let qualifying: Option<u32> = row.take("isQualifying");`
it crashes on null saying it can't convert null to u32. It seems like it should convert null to None?

Thanks for helping me understand

**Comment by blackbeam** [comment]
Hi.

> At what case can it return None if not in the null case

In case if a cell value was already taken — the behavior here is similar to the [`Option::take`](https://doc.rust-lang.org/std/option/enum.Option.html#method.take) — in other words `Row` can roughly be considered as a `Vec<Option<T>>`:

```rust
let row: Row = "SELECT 'cell_value' as `cell_name`".first(&mut conn).await?.expect("a single row will be here");
let cell_value = row.take("cell_name").expect("no panic here because the value of the cell `cell_name` is in the row");
let cell_value2 = row.take("cell").expect("this line will panic because the value of the call `cell_name` was taken at the previous line");
```

> If I try
> 
> ```rust
> let qualifying: Option<u32> = row.take("isQualifying");
> ```
> it crashes on null saying it can't convert null to u32. It seems like it should convert null to None?

For the same reason as described above this should be:

```rust
let qualifying: Option<Option<u32>> = row.take("isQualifying");

// or in case you can assert that the value of `isQualifying` wasn't taken
let qualifying: Option<u32> = row.take("isQualifying").expect("it is absurd if column value isn't there");
```

Consider also:

* [`from_row` function](https://docs.rs/mysql_async/latest/mysql_async/fn.from_row.html) — to convert a row into a tuple
* [`FromRow` derive macro](https://github.com/blackbeam/rust_mysql_common?tab=readme-ov-file#fromrow-derive) — to convert row into a struct
